[IBC] Create initial IBC module (#842)
pokt-network/pocket: ibc/docs/README.md

# IBC <!-- omit in toc -->

- [Overview](#overview)
- [IBC Module](#ibc-module)
  - [Host Configuration](#host-configuration)
- [Components](#components)
  - [ICS-24 Host Requirements](#ics-24-host-requirements)

## Overview

![IBC High Level](./high-level-ibc.png)

> Inter-Blockchain Communication (IBC) is a protocol that facilitates the trustless communication between two chains. This allows them to interact with each other via the relaying of IBC packets. This works by having two IBC enabled chains: **chain A** and **chain B**; both of which will have a light client for the other running on their network. In order for **chain A** to transfer some of their native token to **chain B** a few things must be in place: a connection between the two chains, a channel and port must be opened between them on this connection, a light client for the opposing chain on each end of the connection and finally a relayer is needed to actually transfer the packet. Once all these are in place, **chain A** can commit an IBC packet to its state containing the information around the transaction, and generate a proof detailing its inclusion in the state at the specified height. The relayer will then submit this proof to **chain B** which will verify the proof and if valid **chain B** can then react to this packet in an appropriate fashion (ie. sending an IBC **chain A** token to the specified address on **chain B**).

Reviewer: I asked chatGPT to optimize for readability and clarity and this is what resulted:

```suggestion
Inter-Blockchain Communication (IBC) is a protocol that enables trustless communication between two chains. It allows these chains to interact by relaying IBC packets. The process involves two IBC-enabled chains, referred to as **chain A** and **chain B**, each running a light client for the other chain on their network.

To transfer native tokens from **chain A** to **chain B**, certain prerequisites must be met. First, a connection between the two chains must be established. Then, a channel and port need to be opened on this connection. Additionally, a light client for the opposing chain should be set up on both ends of the connection. Finally, a relayer is required to facilitate the actual transfer of the packet.

Once these components are in place, **chain A** can commit an IBC packet to its state, which contains transaction information. It also generates a proof that specifies the inclusion of the packet in the state at a particular height. The relayer submits this proof to **chain B**, where it is verified. If the proof is valid, **chain B** can respond accordingly, such as by sending an IBC token from **chain A** to the designated address on **chain B**.
```

## IBC Module

Pocket's IBC module is split into numerous components detailed below. The overall module layout is as follows:

**Note:** Not all of the different ICS components have been fully implemented yet, this is a work in progress.

```mermaid
flowchart TB
    subgraph IBC[IBC Module]
        subgraph 23[ICS-23]
            VC[Vector Commitments]
        end
        subgraph 24[ICS-24]
            HR[Host Requirements]
        end
        subgraph I[IBC Interface]
            IF[ICS-25 Handler Interface]
            IR[ICS-26 Routing Module]
        end
    end
    subgraph 2[ICS-02]
        CL[Client Semantics]
    end
    subgraph 3[ICS-03]
        CO[Connection Semantics]
    end
    subgraph 5[ICS-05]
        PA[Port Allocation]
    end
    subgraph 4[ICS-04]
        CP[Channel & Packet Semantics]
    end
    subgraph 20[ICS-20]
        FT[Fungible Token Transfer]
    end
    IBC ---> 2
    IBC ---> 3
    IBC ---> 5
    IBC ---> 4
    IBC ---> 20
```

> ### Host Configuration

Reviewer: Wdyt about "node" instead of "host" here to be consistent with other Pocket docs (and the line below)?

Author: > My observation was that the content under this section seems to be describing how to configure a pocket node to be an IBC host instead.

Yeah I understand now, will edit accordingly

Part of the node configurations relating to the IBC module is as follows:

```json
"ibc": {
    "enabled": bool,
}
```

> If a node enables the IBC module, and is a validator, then the IBC module will automatically create an IBC host on startup. As the host defines the connections, channels and ports - which must stay persistent, the node should be a validator with little risk of suddenly closing any of these while open. As any tokens transferred over a connection/channel/port are unique to that combination they can only be returned over the same combination. If the channel is to close without warning then tokens will be unable to be returned to their source. It is for this reason that only validators are able to become IBC hosts.

Reviewer: Can you say more about what "connections", "channels", and "ports" **are** (i.e. chain state and/or active network endpoint(s) (all?) host(s) need to be listening on)? It's not clear to me given the context, what it would actually mean for a validator to "close" any of them.

> ... any tokens transferred over a connection/channel/port are unique to that combination they can only be returned over the same combination. If the channel is to close without warning then tokens will be unable to be returned to their source. It is for this reason that only validators are able to become IBC hosts.

I feel like this should at least be emboldened - pretty significant detail. :sweat_smile: I love that you've connected it to the rationale behind the aforementioned role constraint. :100:

Author: To my current knowledge these are just entries in the chain state (basically identifiers) and to "Close" a connection for example requires the commitment of a packet to close the connection, the relayer will send this to the other end and the connection will be closed etc. It does not mean any "real P2P" connections are made.

I will update this in the docs

## Components

> The [IBC specification][ibc-spec] details numerous Interchain Standards (ICSs) that together form the IBC protocol. The following gives an overview of the different components implemented in Pocket's IBC module.

Reviewer: Are these double square brackets link placeholders? (also "ICS-24" below)

Author: The links are at the bottom - I assumed I would be reusing certain ones a few times just makes it easier to read while writing

Reviewer: Neat; TIL! I did not not know you could do that :astonished:

### ICS-24 Host Requirements

[ICS-24][ics24] defines the requirements for a host chain to be IBC compatible. This includes the definition of a store system to hold IBC related data in a provable (and also a private) fashion. This implementation uses the [SMT](https://github.com/pokt-network/smt) rather than the IAVL tree used by `cosmos-sdk` for its provable stores. ICS-24 also defines the Event Logging system that is used to store and query IBC related events for the relayers to read packet data and timeouts, as only the proofs of these are stored in the chain state.

> More details on the specifics of the ICS-24 implementation for Pocket can be found [here](./ics24.md).

Reviewer: Nit: I think we should prefer a semantically meaningful link over "here links":

```suggestion
See: [ICS-24](./ics24.md) for more details on the specifics of the ICS-24 implementation for Pocket.
```

# IBC <!-- omit in toc -->

- [Overview](#overview)
- [IBC Module](#ibc-module)
  - [Node Configuration](#node-configuration)
- [Components](#components)
  - [ICS-24 Host Requirements](#ics-24-host-requirements)

> 

Reviewer: Wdyt about adding a persistence (or something) section which covers what network state this module interacts with?

Author: I have added a brief section to cover this but the rest of the persistence info will come in the ICS-24 Store PR as this is where interfacing with persistence actually comes into play

# IBC <!-- omit in toc -->

- [Overview](#overview)
- [IBC Module](#ibc-module)
  - [Node Configuration](#node-configuration)
  - [Persistence](#persistence)
- [Components](#components)
  - [ICS-24 Host Requirements](#ics-24-host-requirements)

## Overview

![IBC High Level](./high-level-ibc.png)

> Inter-Blockchain Communication (IBC) is a protocol that enables trustless communication between two chains. It allows these chains to interact by relaying IBC packets. The process involves two IBC-enabled chains, referred to as **chain A** and **chain B**, each running a light client for the other chain on their network.

Reviewer: Would it make sense to have a "Definitions" or "Glossary" section ([like P2P](https://github.com/pokt-network/pocket/tree/main/p2p#definitions))?

I'm also noticing that both P2P and IBC will be using the term "host", so perhaps both READMEs should also define that term.

> Inter-Blockchain Communication (IBC) is a protocol that enables trustless communication between two chains. It allows these chains to interact by relaying IBC packets. The process involves two IBC-enabled chains, referred to as **chain A** and **chain B**, each running a light client for the other chain on their network.

Reviewer 2: +1 to @bryanchriswhite's suggestion

        end
    end
    subgraph 2[ICS-02]
        CL[Client Semantics]
    end
    subgraph 3[ICS-03]
        CO[Connection Semantics]
    end
    subgraph 5[ICS-05]
        PA[Port Allocation]
    end
    subgraph 4[ICS-04]
        CP[Channel & Packet Semantics]
    end
    subgraph 20[ICS-20]
        FT[Fungible Token Transfer]
    end
    IBC ---> 2
    IBC ---> 3
    IBC ---> 5
    IBC ---> 4
    IBC ---> 20
```

### Node Configuration

Part of the node configurations relating to the IBC module is as follows:

```json
"ibc": {
    "enabled": bool,
}
```

If a node enables the IBC module, and is a validator, then the IBC module will automatically create an IBC host on startup. As the host defines the connections, channels and ports - which must stay persistent, the node should be a validator with little risk of suddenly closing any of these while open. Any tokens transferred over a connection/channel/port are unique to that combination they can only be returned over the same combination.

**If the channel is to close without warning then tokens will be unable to be returned to their source. It is for this reason that only validators are able to become IBC hosts.**

> _Note_: Connections, Channels and Ports in IBC are not the same as networking connections, channels and ports. They are stored in the chain state and are used by relayers to signify where each IBC packet should go when being relayed. When closing a channel the IBC host must submit to the state a `ChanCloseInit` IBC packet. If this happens without warning, the funds transferred on this channel will become unrecoverable.

Author: We could build in a local cache that holds packets yet to be reaped from the mempool and included in the state in #854 this could allow for local changes to be persistent. Note the Event logging system will also hold this information but if not commited to state will fail to be verified when relayed.
